### ISSUE TITLE
vNext (1.6) Release Discussion

### ISSUE THREAD

**Issue by jzoss** [jzoss/Git-Source-Control-Provider#30]
Thread to talk about the next release and nightly builds.. 

So talk about what you might want to see in the next release, status of nightly builds, bug you may have found the a early release. 

1. Develop new ideas here, sure you can make a new issue and label it an enhancement  but feel free to brainstorm here.

2. Ideas on how to make the code better.. Check out the code.. look around find a better way of doing something.. talk about it, fork it .. all that fun stuff. 

3. Talk about merging projects.. Should we merge this project with Git Tools.. should Git Tools merge into this project? Do we still want to merge it it has to be a SCC plugin? Tell me what you think.

**Comment by jzoss** [comment]
For more info on point 3  https://github.com/jzoss/Git-Source-Control-Provider/issues/13#issuecomment-180520688

**Comment by jbestrom** [comment]
@jzoss I would say just from my OCD point of view is to remove all unneeded comment blocks. I noticed quite a few and if it is all dead code now would be the time to remove it.

Also a couple places where there are some stopwatch stuff (I assume it was to track how long things took)

**Comment by jbestrom** [comment]
@jzoss As for porting to VS2013 what kind of work would that entail?

I think it could help others if they are stuck there in VS2013 and use git.

**Comment by jzoss** [comment]
@jbestrom Totally agree about the dead code.. I wanted to clean it up before the release, just wanted to get the thing released.. Anyone know of tools that search for dead code? 

As for vs2013 ..  Might be as simple as a config update.. Not sure

**Comment by jzoss** [comment]
So.. @jbestrom,  @PureKrome, @teebee76, @0x084E - Wondering how 1.5 is going.. and if you have had any issues.. I have 2 from the VSgallery - "Double-click on a file in Pending Changes instead of opening the file I get an error dialog that says "The method or operation is not implemented" and someoen who can't save files anymore.. Just wondering how it's all going.. This thread is so quiet.. people must either really hate it, or are happy :) @teebee76 You said you had feature suggestions.. you should.. suggest :)

**Comment by jbestrom** [comment]
I for one personally hate it! /sarcasmoff      For real though great job!

I haven't been in VS2k15 much the past week due to those resharper issues I mentioned before.

The double-click in the pending changes I haven't noticed an issue when I am in VS2k15. I do remember you fixed an issue I reported where if you didn't have an actual solution selected it would toss an error https://github.com/jzoss/Git-Source-Control-Provider/issues/29#issuecomment-175134053 so not sure if it is something similar to that.

I'm going to see about changing the config files to see if I can get it to run in VS2k13 locally as i'm in that world for the time being. I'll let you know if I notice any issues there.

**Comment by teebee76** [comment]
Not noticed anything unusual until now :D (since you mentioned the double click bug, I just had to go in an check for myself. 

No problem for me with when double-clicking the file under changed files. The file opens (or focuses) as it should. Taking me either to the first line, if it opens, or to the last visited line if it focuses. So just perfect.

However I tried double-clicking a changed block in the changes panel, and ofcourse this is not implemented so it justed marked the word I was hovering over. .. And yet again I managed to find a totally unimportent bug/strange behaviour. 
Don't really know if other programs act like this too, just was really baffled when I saw it, and though .. that was strange .. 
allright enough of the suspense ..  If I mark a word in the middle of an added or deleted line (the color-coded ones)  the rest of the line loses its colorcoding .. I'll show an example to clearify 
unmarked:
![image](https://cloud.githubusercontent.com/assets/5390116/12941978/bc00830e-cfd7-11e5-9b66-1a71c93f6e0e.png)
marked:
![image](https://cloud.githubusercontent.com/assets/5390116/12941896/34a13ab6-cfd7-11e5-9447-b9e67f7a6b4d.png)
looks like maybe the background-color is set in the start of the line, and not reset if part of the line is marked...

When it comes to branching and checkins and the other important stuff of the plugin, It all works great for me. Though I have been truly sloppy with my checkins this last week, so I haven't really had time to test a lot of itterations.

**Comment by jzoss** [comment]
@jbestrom did you find where you can config that value for 2013? I can help if you need. I don't want ot add it officially just yet, need to figure out #13 first.

**Comment by jbestrom** [comment]
@jzoss  I did get it installed but got the error below

![image](https://cloud.githubusercontent.com/assets/16417291/12954938/f53ab6b6-cfdd-11e5-910c-1f2876e7ad53.png)

Here is the only stuff I changed to get it installed in source.extension.vsixmanifest.

`<Installation>`
`-    <InstallationTarget Version="[14.0,15.0)" Id="Microsoft.VisualStudio.Pro" />`
`-    <InstallationTarget Version="[14.0,15.0)" Id="Microsoft.VisualStudio.Enterprise" />`
`-    <InstallationTarget Version="[14.0,15.0)" Id="Microsoft.VisualStudio.Community" />`
`+    <InstallationTarget Version="[12.0,15.0)" Id="Microsoft.VisualStudio.Pro" />`
`+    <InstallationTarget Version="[12.0,15.0)" Id="Microsoft.VisualStudio.Enterprise" />`
`+    <InstallationTarget Version="[12.0,15.0)" Id="Microsoft.VisualStudio.Community" />`
`   </Installation>`
`   <Dependencies>`
`     <Dependency Id="Microsoft.Framework.NDP" DisplayName="Microsoft .NET Framework" d:Source="Manual" Version="[4.5,)" />`
`-    <Dependency Id="Microsoft.VisualStudio.MPF.14.0" DisplayName="Visual Studio MPF 14.0" d:Source="Installed" Version="[14.0]" />`
`+    <Dependency Id="Microsoft.VisualStudio.MPF.12.0" DisplayName="Visual Studio MPF 12.0" d:Source="Installed" Version="[12.0]" />`
`   </Dependencies>`

If I missed something let me know please. Thanks

**Comment by jzoss** [comment]
@jbestrom Ok .. took a quick look at lunch.. I think it might be a wee bit more complicated than I thought .. I really went all VS 2015 with this.. So.. I would take a guess.. But.. I used nuget to get all the visual studio extension dll.. Try swapping out Shell.14.0 for Shell.12 in the GitSccProject .. I can try to take a closer look later as well.

**Comment by jbestrom** [comment]
@jzoss Your zip file seems to be working.

Could you add a patch file here or just tell me what to change so when I pull down the code and build locally I can reapply them for VS2k13.

**Comment by jzoss** [comment]
I started by doing what you did, then I Removed the nuget for Microsoft.VisualStudio.Shell.14.0 and installed VSSDK.Shell.12 .. I will look into if there are disadvantages just keeping it at the lower version...

**Comment by jbestrom** [comment]
@jzoss don't waste too much time I just upgraded to ReSharper 10.1.EAP.1 and it seems to be working a little better so I may be able to fully make the jump to VS2K15.

**Comment by jbestrom** [comment]
@jzoss Haven't figured out when/why it is happening but the "Show only files in solution" in the pending changes window isn't working 100% of the time.

I'll do my best to figure out when/why it is happening but I just wanted to give you a heads up.

**Comment by jzoss** [comment]
@jbestrom Thanks.. Yea.. I would not be surprised if there was a bug somewhere.. I am going to do  1.5 R2 to fix a few little bugs here and there.. maybe early next week.. So if you have any idea, on have anyway I can recreate it .. that would be cool

**Comment by PureKrome** [comment]
IMO
- Create a milestone called *1.5.1* (Follow SemVer please)
- Add issues _per_ bug/problem and tag then to that milestone.
- :money_with_wings:

**Comment by jzoss** [comment]
@PureKrome  Are you sure I should just increment the major version all the time :) Like chrome or firefox :) 

Anyway.. I was thinking of updating the C.I. version to get better build number.. Question.. @PureKrome would you be offended by the next on being 1.5 R2 ? I'm not sure if I like that yet.

**Comment by PureKrome** [comment]
TL;DR;
- follow SemVer _always_
- next version should be **1.5.1**

---

> would you be offended
(In a very polite and respectable voice) : Yes, I would be.

The now-and-future is SemVer IMO. It just makes things SOOO much easier.

Assumption: you know what SemVer .

What you are doing here is a bug fix to the current release. It's not a major breaking change in logic or api methods. It's not some new functionality or a serious bug fix. It's "_a few little bugs here and there_". These are minor improvements.

None the less, don't be scare of bumping the major or minor numbers. They are there for a purpose.

But most importantly, the bumping of _any_ of the numbers accurately explain the reason for a new release/vesion. A minor number bumped? kewl, _some_ minor bugs or stuff has been changed/added. 

This concept of :hankey: fancy pancy verbose descriptions of crap that comes out of MS in their f'ing branding/marketing dept is a insult to us all. an *R2*? A waa? wtf does a Release 2 _mean_? oops, we shouldn't have release the first one .. forget about that.. this is the one you want....

pah-lease.

Marketing needs to leave us dev's alone. Stick to awesome SemVer. Embrace it.

(and it's easy to do with AppVeyor for your CI, btw).

<end rant>

**Comment by jzoss** [comment]
Yea right now the build server adds to the patch each build .. So since I've checked in a few times it is 1.5.7 so.. I would want to change that to either 1.5.0.x or make the next build 1.6 .. I guess I will figure that out :)

**Comment by PureKrome** [comment]
I'm now going to go into a wee bit more advanced stuff/suggestion:

- create 2 branches: `master` and `dev`
- `master` builds and is a normal formal release. (bonus points: auto-publishes to VS marketplace thingy)
- `dev` has a _different_ version number: `1.6.0.<build-number>-dev`. Publish to a GH release.

what's important here is that we're still following SemVer and that having some text (in this case, `-dev`) appended to the version means that this product is a _pre-release_ product.

so now, you can keep pushing stuff up to the `dev` branch as much as possible. when ready, you Pull Request from `dev` -> `master` and a NEW version is created.

Using AppVeyor for your CI can do most (all?) of this.

ASSUMPTION 1: it's possible to create a VS Extension with a version that has some text in them. eg. `1.6.0.123232-dev`
ASSUMPTION 2: it's possiible to use powershell to publish a compiled VS extension to the VS marketplace thingy.

^ the above concept is basically what I do with my public OSS nuget library/packages. (I can provide examples, too).

**Comment by jzoss** [comment]
@PureKrome Ok.. I've been a little busy.. at home at and work.. BUT I finally got versioning sorted.. The C.I. Build now goes to the release tab in git.. It even cleans itself up :) It was harder than it shoudl have been.

You can't deploy automatically and adding -dev to the version makes visual studio cranky ..   

Anyway.. dev version goes .. 1.5.0.(1...2..3..4...) and the release version goes like 1.5.(1,2,3) This way you can feel free to install the dev version and it will autoupgrade after the next release. the dev version for the next release will go 1.5.1.(1,2,3 etc) 

Anyway.. please all check out the latest dev release.. has a bunch of bug fixes. @teebee76 Tell me if you think  #37 is acting better.

**Comment by PureKrome** [comment]
shame about VS getting cranky with `-dev` (or any postpended text) in the version :/

**Comment by teebee76** [comment]
Will download now and keep an eye out for glyph-irregularities :)

**Comment by jzoss** [comment]
@teebee76 Yea.. I have 2 builds on appveyor now.. here is the dev build https://ci.appveyor.com/project/jzoss/git-source-control-provider-bfftg/build/artifacts

**Comment by teebee76** [comment]
Yeah I know,  after I wrote it,  I figured I just was on the wrong branch when checking the readme... 
Didn't think i actually posted that comment,  but obviously I did...

**Comment by jzoss** [comment]
HI all .. work on 1.6 has been a little slow.. as I have been a bit busy..and I spent some time on the release automation. However.. if you look the release at the step 1 is just about complete.. Release 1.5.1. the latest C.I. build includes a bunch of bug fixes and the official upgrade to the new libgit2sharp. Please install and tell me what you think before I release.. Anyway.. after this back to 1.6 :)

**Comment by jzoss** [comment]
Hey sorry.. Just noticed there was a bug in my deployment..  Anyway please check out the latest 1.5 dev build in the release tab.  I plan on making it the official 1.5.1 release.

**Comment by jzoss** [comment]
@teebee76 @PureKrome @teebee76 - Just wondering if any of you have tried version 1.5.0.52.. Seems solid to me.. I was going to promote it to 1.5.1 but I just wanted to see if others had been using it. 

Starting to plan out 1.6 a bit.. I've been much more busy at work and at home some my time has gone down.. but I'm making some progress.

**Comment by teebee76** [comment]
It appears I'm still on 1.5.0.50, so I will get the latest and check it out. I noticed some strange behaviour earlier today, but I will see if I can recreate in the most up-to-date version before posting.

**Comment by ddobrev** [comment]
Please consider https://github.com/jzoss/Git-Source-Control-Provider/issues/13#issuecomment-210268850 as my consideration about 1.6.

**Comment by jzoss** [comment]
@teebee76 @jbestrom @PureKrome @ddobrev  Hey everyone... it's been a while.. So.. it's that time again.. I have a pretty stable release ready for you all. That is .. it will probably become the official release once I update all the release notes.. But I would very much like some outside testing.. This release I focused a lot on performance, so look at that.. Also I updated the scc provider for the Update 2.. So that means some cool status bar updates.. you can see status on the status bar.. 

![statusbar](https://cloud.githubusercontent.com/assets/3586254/15159754/d5b40796-16bb-11e6-97bb-25ecdd6f42ef.png)

You can open the pending changes by clicking on the changed file count.. 

![openpending changes](https://cloud.githubusercontent.com/assets/3586254/15159774/e9f1cb80-16bb-11e6-855b-04402216d811.png)

and even switch between repos if you have multiple repos 

![mult-repo](https://cloud.githubusercontent.com/assets/3586254/15159791/fcb305fe-16bb-11e6-9501-e10df4ba4aa0.png)


I also readded the ability to click on a change in the diff window and open up the file on the line that changed. So once again... I spent way too much time on this... so I hope you like it.. I think this is the first external plugin to use the new status bar abilities in update 2 .. so I hope you like it.

Also.. as this may be the 1.6 release.. I was planning on moving the "What to do in 1.7" to the gitter chat room, what does everyone think? 

Anyway get the latest C.I. build from the release tab to try it out.. help me test. Someone besides me.. if you could test init a new repo from the status bar that would be great

**Comment by ddobrev** [comment]
@jzoss I have been using the build from two weeks ago - when we were discussing performance at https://github.com/jzoss/Git-Source-Control-Provider/issues/46 - without updating. I haven't encountered any bugs, first, and I stopped having CPU pressure at each change. So as far as I am concerned, you might as well go ahead and make an official release. I have installed the latest build just in case but I doubt I'd have any trouble.

**Comment by jbestrom** [comment]
@jzoss I'm finally able to consistently run VS2k15 now so I'll make sure to keep an eye out for any issues I see and report back on them.

**Comment by PureKrome** [comment]
I've been using the latest version since well .. 7 days ago from this comment.

no probs so far. 

NitPick: when i hover the mouse cursor over the pencil icon in the `status bar`, i get a tooltip that is .. well.. empty.

**Comment by jzoss** [comment]
@PureKrome...  Dam you!  I totally forgot to add the tool tip.. I was hoping no one would notice 😀.  I'll fix it in the next release.

**Comment by PureKrome** [comment]
First thing I did after installation - ooOOO! new shiney bits .. what do they _all_ do?

**Comment by jzoss** [comment]
@PureKrome .. You win.. 1.6.3 is out... with tooltips :)

**Comment by jzoss** [comment]
@PureKrome , @jbestrom, @teebee76 , @ddobrev - I was hoping you guys could do a sanity test for me.. I have a new C.I. build [here](https://github.com/jzoss/Git-Source-Control-Provider/releases/tag/GSCP-CI) .. I should be exactly the same as the last release build, except 65% smaller. I just cleaned up the build and stuff.. Anyway.. if you could install it and make sure you have no issues, that would rock.

**Comment by jzoss** [comment]
@PureKrome , @jbestrom, @teebee76 , @ddobrev - I was hoping you guys could do a sanity test for me.. I have a new C.I. build [here](https://github.com/jzoss/Git-Source-Control-Provider/releases/tag/GSCP-CI) .. I should be exactly the same as the last release build, except 65% smaller. I just cleaned up the build and stuff.. Anyway.. if you could install it and make sure you have no issues, that would rock.

**Comment by PureKrome** [comment]
8mb down to just-under 3mb....

**Comment by 0x084E** [comment]
intsalled this morgning, working well so far :+1:

**Comment by jzoss** [comment]
@0x084E , @PureKrome .. Thanks for testing.. should be no issues at all :) I need to fix at least 1 bug.. For some reason I think it may not be a good idea to do a new release just because the download is smaller...

**Comment by PureKrome** [comment]
+1 on _not_ releasing because of a new installer (vs bugfixes/features).

**Comment by ddobrev** [comment]
I've been using the small for two weeks, no problems.

**Comment by PureKrome** [comment]
Same here - been using it since it was released. all good so far...

**Comment by jzoss** [comment]
Thanks Everyone.. You should have gotten the auto-update to 1.6.4 :)  Thanks for testing. I like it being way smaller.

**Comment by jzoss** [comment]
@jbestrom @teebee76 @PureKrome @ddobrev @0x084E .. HI to all the cool people.. It's been a while.. I got a new job, so things slowed down a bit..

Anyway there is a new c.i. build and it could become 1.6.5. But I want a little feedback first. 

1. I think before 1.7 I need a new name. 1.7 is going to hopefully run in the new visual studio. So I was thinking Git SCC Plus or something.

2. Check out the new features.. the big one is it will auto-add files to a project. Also.. try adding a new project to a solution then right clicking on the project and going to scc menu, you should be able to add that project to to source control. Not sure if I like that method yet, but I am not sure a better way to add projects.,

3. Options have changed.. they used to store options in your user directory, and that is bad, so it is now in the registry. Also the options screen should look better on high res screens.

Anyway if you could take it for a spin for me . that would be great.

**Comment by ddobrev** [comment]
1. I think the current name is fine, it clearly describes what the add-in does. Other than that, I have no particular suggestions;
2. Thank you for this one, as I mentioned when I added the feature request, this was the only major inconvenience with the original project. About adding projects, is there a problem with just going through each added file and adding it to version control automatically?
3. I strongly disapprove of this change. The registry is responsible for at least 80 % of all problems with Windows so if there's a way not to use it, it's always preferable. In particular, I think Git SCC will fall into the regular trap of applications not cleaning their registry entries when they are removed. I was recently cleaning up my registry and found records for Skype 3.8 which I last used 4 years ago! Even if you've taken care of this problem, the very way the registry works will have Git SCC affect the whole system. The whole registry is read each time the system starts up regardless of which applications are actually used. I personally am satisfied with Git SCC now and even though I will miss the automatically added files a little, I won't upgrade to the new version if it uses the registry.

**Comment by jbestrom** [comment]
1. Don't care about the name as long as I can find it :)
2. I'll test when I get a chance.
3. Why do you think that storing in the user directory is bad? I semi agree with @ddobrev about the registry being a bad choice.

**Comment by jzoss** [comment]
@ddobrev 
To answer #2 No.. Once the project is added it adds the files.. and once you right click on the project ad add it goes through and add all the files and adds them. The issue is it's hard to tell when to add the project... The project says it's loaded before it sorta is.. if that makes sense.. So when you look for all the files they might not be loaded.. Anyway I may work on that.

As for #3 : I am now saving the options the standard way all other extensions save options..  There is a built in user settings store. It happens to in the back end store them in a the registry, but I am not saving them directly in there.. But if people don't like it I can switch back.. but the rest of your extensions are save setting in there already.

**Comment by jbestrom** [comment]
I'm cool with Q3 change then I was taking it that you were creating new registry entries just for this extension. Does the built in user store handle the removing correctly when you remove the extension?

**Comment by ddobrev** [comment]
@jzoss 
2. I understand, it's OK. Adding entire new projects happens way too rarely given the entire lifetime of developing a program so we'll be just as fine without;
3. I myself would be tempted by going with the standard. However, when the standard is bad, as it often is with Microsoft, the custom solution is preferable. So I support @jbestrom 's suggestion of you sharing what the problem with the user directory is, we could perhaps discuss is and think of a better solution.

**Comment by jzoss** [comment]
No big reason to make the change.. just didn't like the file there, it never gets cleaned up, and I was trying to make the options follow you with the login.

**Comment by ddobrev** [comment]
I see, in this case I think the old way is better. You most probably have some hooks when the add-in is removed where you can delete the file if you want. But even if it remains, it's not a big deal because it's extremely small (how large can a text file be?) and it's on the file system. Besides, this way settings persist when manually removing and then reinstalling the add-in (for example, when testing preliminary versions).

**Comment by PureKrome** [comment]
What are the *options* that are getting saved into a registry?

Personally - i think the registry is bane of all evil. A simple settings file side-by-side the app or in some user folder (for user based settings) make way more sense to me.

**Comment by 0x084E** [comment]
Generally I'm with you people regardig the registry, but I think this case is different.

Settings are not stored direcly in the registry but as @jzoss mentioned in the user settings store which is obviously stored in the registry but has one advantage above a custom file (and @jzoss already mentioned):
Settings are synchronized with your profile. Also settings are exported with the standard VS export settings functionatlity and can be imported back likewise.
So I think one should use what VS gives you to integrate your extension like it was intended.

P.S.:
If you followed the VS blog regarding the upcoming version you might have noticed that they are in progress of reducing / removing the registry access in VS to make it more portable (source: https://blogs.msdn.microsoft.com/visualstudio/2016/04/25/anatomy-of-a-low-impact-visual-studio-install)
So probably this discussion will be obsolete with the next VS version

**Comment by jzoss** [comment]
I am going to give the win to @0x084E :) for doing the research.  Now it seems the best move to use the user setting store and let it do it's thing. Next version of vs looks to fix it, and if I would have just said changed the options to use the standard user store, I don't think anyone would have cared :) Anyone disagree?

**Comment by jbestrom** [comment]
Wouldn't have even noticed :) you used the dreaded "Registry" word though.

**Comment by jzoss** [comment]
@jbestrom I know.. I should have not used the "R" word..

**Comment by PureKrome** [comment]
Sounds great!

**Comment by jzoss** [comment]
Hi all again.. So.. I just pushed an new c.i. build... So... @ddobrev It now adds projects automatically, just check the options..

Anyway, I'd like people to test it out hard core.. Doing the auto-add exposed some long simmering bugs, that under normal usage may not have shown up.. Anyway, make sure all the file status stuff still works right.. Performance to me seems a BIT better, but make sure you don't think it is worse.. 

Last question - when a project is auto-added should we have a popup come up and warn you or allow you to not add it?

**Comment by ddobrev** [comment]
I don't think so, that is - why would you not want to add it? Of course, VS does generate useless Class1.cs files by default but it's none of GitSCC's responsibility.

**Comment by jzoss** [comment]
Thanks again everyone .. I just pushed the final 1.6.5(.1) release..  Could not do it with out you guys!

**Comment by jzoss** [comment]
Hey all.. I wanted to point out that the next release is almost here.. @jbestrom @teebee76 @ddobrev @0x084E @PureKrome .. and anyone else..I would love for you test.. https://github.com/jzoss/Git-Source-Control-Provider/releases/tag/GSCP-CI

This includes Vs 2017 support.

**Comment by jbestrom** [comment]
Been on the 2017 version since you updated the code the first time and have not noticed any issues for what I do. I don't do a lot of branching or committing through the extension though as I have to send everything to an SVN server so I have to us the git svn dcommit/rebase commands.

**Comment by PureKrome** [comment]
Installed. Will report back if any probs/issues/etc.

**Comment by jzoss** [comment]
@jbestrom @teebee76 @ddobrev @0x084E @PureKrome - Hi party people.. I released 1.7.1 today.. I has a few super minor bug fixes and support for VS 2017. Yell if you run into issues ..
